include/dcsam/DCSAM_utils.h

```diff
@@ -17,6 +17,23 @@
 
 namespace dcsam {
 
+/*
+ * Numerically stable log-sum-exp function.
+ */
+inline double logSumExp(const std::vector<double> &values) {
+  double max = -std::numeric_limits<double>::infinity();
+  for (size_t i = 0; i < values.size(); i++) {
+    if (values[i] > max) {
+      max = values[i];
+    }
+  }
```

Comment:
not sure whether using C++ internal functions (e.g. max element, sum, etc.) will improve speed here

Reply:
https://stackoverflow.com/a/45960829 From a quick google search